src/meta/src/barrier/mod.rs

```diff
@@ -910,11 +913,101 @@ impl GlobalBarrierManager {
             let fail_nodes = complete_nodes
                 .drain(index..)
                 .chain(checkpoint_control.barrier_failed().into_iter());
+            tracing::warn!("Failed to commit epoch {}: {:?}", prev_epoch, err);
             self.failure_recovery(err, fail_nodes, state, checkpoint_control)
                 .await;
         }
     }
 
+    async fn recover_mview_progress(&self) -> MetaResult<()> {
+        let creating_tables = self.catalog_manager.list_creating_mviews().await;
+        let creating_table_ids = creating_tables
+            .iter()
+            .map(|t| TableId { table_id: t.id })
+            .collect_vec();
+
+        let mut senders = HashMap::new();
+        let mut receivers = Vec::new();
+        for table_id in creating_table_ids.iter().copied() {
+            let (finished_tx, finished_rx) = oneshot::channel();
+            senders.insert(
+                table_id,
+                Notifier {
+                    finished: Some(finished_tx),
+                    ..Default::default()
+                },
+            );
+
+            let fragments = self
+                .fragment_manager
+                .select_table_fragments_by_table_id(&table_id)
+                .await?;
+            let internal_table_ids = fragments.internal_table_ids();
+            let internal_tables = self.catalog_manager.get_tables(&internal_table_ids).await;
+            let table = self.catalog_manager.get_tables(&[table_id.table_id]).await;
+            assert_eq!(table.len(), 1, "should only have 1 materialized table");
+            let table = table.into_iter().next().unwrap();
+            receivers.push((table, internal_tables, finished_rx));
+        }
+
+        let table_map = self
+            .fragment_manager
+            .get_table_id_actor_mapping(&creating_table_ids)
+            .await;
+        let upstream_mv_counts = self
+            .fragment_manager
+            .get_upstream_relation_counts(&creating_table_ids)
+            .await;
+        let definitions: HashMap<_, _> = creating_tables
+            .into_iter()
+            .map(|t| (TableId { table_id: t.id }, t.definition))
+            .collect();
+        let version_stats = self.hummock_manager.get_version_stats().await;
+        // If failed, enter recovery mode.
+        {
+            let mut tracker = self.tracker.lock().await;
+            *tracker = CreateMviewProgressTracker::recover(
+                table_map,
+                upstream_mv_counts,
+                definitions,
+                version_stats,
+                senders,
+            );
+        }
+        for (table, internal_tables, finished) in receivers {
+            let catalog_manager = self.catalog_manager.clone();
+            tokio::spawn(async move {
```

Comment:
Should we manually abort these background receiver tasks for each recovery attempts? Will they interfer with each other?

Reply:
Good catch. I guess not, the previous finish receiver will get an channel close error because the tracker is dropped. It seems like that only `recover_mview_progress`at the beginning of the recovery is already enough. 🤔

Reply:
Not quite sure what you mean. Meta doesn't manually abort any stream jobs here.
Unless you're referring to the user having to manually cancel the stream jobs?